models.py

```diff
@@ -247,8 +274,17 @@ def __init__(self, n, emsize, vocab_size, normalization,
               residual_list(conv_proto, n//2-1, normalization)))
 
         self.postfix = nn.Sequential(
-                *(residual_list(conv_proto, n//2, normalization,
+                *(residual_list(conv_proto, n//2,
+                                normalization=normalization,
                                 last_relu=(output_embeddings_init is not None))))
+        if use_external_batch_norm:
+            assert external_batch_max_r is not None, \
+                    'to use external batchnorm in reccurent part' + \
+                    ' you need to specify maximum number of recurrent steps'
+            self.rec_batchnorm_list = nn.ModuleList([
+                nn.BatchNorm1d(emsize, momentum=0.1) for _ in range(2 * (n // 2) * external_batch_max_r)])
```

Comment:
W obecnej wersji to nie bierze pod uwagę np. tego, że do pierwszego batchnorma wchodzą wszystkie dane (niezależnie od liczby obrotów rekurencji w enkoderze).

Reply:
To nie szkodzi; agregujemy statystyki dla każdej długości L osobno, czyli ze stałą ilością obrotów rekurencji.

Reply:
Czyli, jeżeli dobrze rozumiem, zmieniamy trochę semantykę `external batch norm`? Teraz używamy tego (tylko?) w kontekście przeliczania statystyk przy ewaluacji?

Reply:
Dokładnie tak. Podczas treningu i tak nie ma żadnej różnicy, czy masz jeden BN, czy kwadratowo wiele.